### PR TITLE
Multiactions/linucb

### PR DESCRIPTION
**Pull request by taweihuang**
Sorry for this super large PR. Since the basebandit changes, the circle CI cannot pass until all bandit algorithms are updated.

### `get_action(context, n_action)`:

(1) Parameters:
- `context`: For UCB1 and EXP3, `context` is `None`; for LinUCB and Thompson Sampling with linear payoff, `context` is a dictionary of the form `{action_id: context_vector}`. For EXP4P, `context` is a dictionary of the form `{expert_id: advice_vectors}`.
- `n_action`: int, number of actions your want the bandit to recommend.

(2) Return:
-  history_id: int, The history id of this query result.
- action_recommend : list of dictionaries. In each dictionary, it will contains {Action object, estimated_reward, uncertainty}.

### `reward(history_id, reward)`:

(1) Parameters:
- history_id : int, the history id of the action to reward.
- reward : dictionary of the form `{action_id: reward}`, where reward is a float

### `add_action(actions)`:
(1) Parameters:
- actions : list of `Action` objects for recommendation

## Refactor Exp4P

In the original algorithms, we require "models" (i.e. a list of scikit-learn supervised learning models) as one parameter for initialization. However, I think it is not quite flexible for users  since they may use models. The new version set `context` as a dictionary of expert advice experts, which is of the form `{expert_id: advice_vectors}`.

## Other Minor Changes

Due to the change of bandit algorithms, followings are also changes.

(1) Update the docstrings.
(2) Update all tests for bandit algorithms.
(3) Update all simulation files.
(4) Update the MovieLens examples.

### CODE REVIEW

File: striatum/bandit/bandit.py [hunk 1]
```diff
@@ -4,43 +4,60 @@
...
+        title: string
+            The title of this action.
+
+        content: object
+            The content of this action.
+    """
+    def __init__(self, action_id, title, content):
+        self.action_id = action_id
+        self.title = title
+        self.content = content
```
Comment (ianlini) [inline striatum/bandit/bandit.py]: @stegben

File: striatum/bandit/bandit.py [hunk 2]
```diff
@@ -4,43 +4,60 @@
...
 
-    modelstorage : modelstorage object
-        The modelstorage object to store model parameters.
+        actions : list of Action objects
+            List of actions to be chosen from.
 
-    actions : list of Action objects
-        List of actions to be chosen from.
+        actions_id: list of integers
+            List of all action_id's.
```
Comment (ianlini) [inline striatum/bandit/bandit.py]: Where are the underscores? 
`action_ids` may be better

File: striatum/bandit/bandit.py [hunk 3]
```diff
@@ -4,43 +4,60 @@
...
     """
+
     def __init__(self, historystorage, modelstorage, actions):
         self._historystorage = historystorage
         self._modelstorage = modelstorage
         self._actions = actions
+        self._actions_id = [actions[i].action_id for i in range(len(actions))]
```
Comment (ianlini) [inline striatum/bandit/bandit.py]: It's not a good idea to have a variable in two places

File: striatum/bandit/bandit.py [hunk 4]
```diff
@@ -58,35 +75,48 @@ def actions(self):
...
         """Return the action to perform
 
-        Parameters
-        ----------
-        context : {array-like, None}
-            The context of current state, None if no context avaliable.
+            Parameters
+            ----------
+            context : dictionary
+                Contexts {action_id: context} of different actions.
```
Comment (ianlini) [inline striatum/bandit/bandit.py]: why indent?

File: striatum/bandit/bandit.py [hunk 5]
```diff
@@ -58,35 +75,48 @@ def actions(self):
         return self._actions
 
     @abstractmethod
-    def get_action(self, context):
+    def get_action(self, context, n_action=1):
```
Comment (ianlini) [inline striatum/bandit/bandit.py]: `n_actions`

File: striatum/bandit/bandit.py [hunk 6]
```diff
@@ -58,35 +75,48 @@ def actions(self):
...
+                Number of actions wanted to recommend users.
 
-        action : Actions object
-            The action to perform.
+            Returns
+            -------
+            history_id : int
+                The history id of the action.
+
+            action_recommend : list of dictionaries
```
Comment (ianlini) [inline striatum/bandit/bandit.py]: use a noun

File: striatum/bandit/bandit.py [hunk 7]
```diff
@@ -58,35 +75,48 @@ def actions(self):
...
-        Parameters
-        ----------
-        history_id : int
-            The history id of the action to reward.
+            Parameters
+            ----------
+            history_id : int
+                The history id of the action to reward.
+
+            reward : dictionary
```
Comment (ianlini) [inline striatum/bandit/bandit.py]: `rewards`?

File: striatum/bandit/bandit.py [hunk 8]
```diff
@@ -58,35 +75,48 @@ def actions(self):
...
+    @abstractmethod
+    def add_action(self, actions):
+        """ Add new actions (if needed).
 
-        reward : float
-            A float representing the feedback given to the action, the higher
-            the better.
+            Parameters
+            ----------
+            actions : list
```
Comment (ianlini) [inline striatum/bandit/bandit.py]: `list` -> `iterable`

File: striatum/bandit/bandit.py (hunk 3 above)
Comment (stegben) [inline striatum/bandit/bandit.py]: Instead of creating another variable, a possible design to extract what you want is to add a method.
```python
def get_action_ids(self):
    return [self._actions[i].action_id for i in range(len(self._actions))]
```

File: striatum/bandit/bandit.py (hunk 4 above)
Comment (stegben) [inline striatum/bandit/bandit.py]: I think this issue is quite small. If the docstring type is consistent everywhere, that's OK.

File: striatum/bandit/bandit.py (hunk 3 above)
Comment (ianlini) [inline striatum/bandit/bandit.py]: or
```
@property
def action_ids(self):
    return (self._actions[i].action_id for i in range(len(self._actions)))
```
depends on how you use it.

File: striatum/bandit/bandit.py (hunk 4 above)
Comment (ianlini) [inline striatum/bandit/bandit.py]: No. We are following a convention, and he is changing the convention without reason. 
(That's not good to have so many deep indent.)

Comment (stegben) [inline striatum/bandit/bandit.py]: @taweihuang maybe you would like to reference PEP 257

Comment (ianlini) [inline striatum/bandit/bandit.py]: we actually refer to sklearn's source code

File: striatum/bandit/bandit.py (hunk 1 above)
Comment (stegben) [inline striatum/bandit/bandit.py]: I suggest to create a abstract method since for different application, action may have different format.

Comment (taweihuang) [inline striatum/bandit/bandit.py]: How about retain `self.action_id`? It's the most important part of an `Action` object for contextual bandit, and adding this can avoid users' forgetting giving `action_id`.

**Comment by ianlini** [comment]
@stegben any other problems? 
I still don't have time to go through all algorithms' code.

**Comment by ianlini** [comment]
I hope that this branch will be merged today

**Comment by stegben** [comment]
LGTM
